scilpy/image/operations.py

```diff
@@ -112,396 +113,463 @@ def _validate_float(x):
+    if len(input_list[0].header.get_data_shape()) > 3:
+        return np.average(input_list[0].get_fdata(dtype=np.float64), axis=-1)
+    else:
+        return addition(input_list, ref_img) / len(input_list)
 
 
-def std(input_list):
+def std(input_list, ref_img):
     """
     std: IMGs
         Compute the standard deviation average of multiple images.
         If a single 4D image is provided, compute the STD along the last
         dimension.
     """
     _validate_length(input_list, 1, at_least=True)
-    _validate_arrays(*input_list)
-    ref_array = input_list[0]
+    _validate_imgs(*input_list, ref_img)
 
-    if len(input_list) == 1 and not ref_array.ndim > 3:
+    if len(input_list) == 1 \
+            and not len(input_list[0].header.get_data_shape()) > 3:
         logging.error('This operation with only one operand requires 4D data.')
         raise ValueError
 
-    in_data = np.squeeze(np.rollaxis(np.array(input_list), 0,
-                                     input_list[0].ndim+1))
-
-    return np.std(in_data, axis=-1)
+    if len(input_list[0].header.get_data_shape()) > 3:
+        return np.std(input_list[0].get_fdata(dtype=np.float64), axis=-1)
+    else:
+        mean_data = mean(input_list, ref_img)
+        output_data = np.zeros(input_list[0].header.get_data_shape())
+        for img in input_list:
+            if isinstance(img, nib.Nifti1Image):
+                data = img.get_fdata(dtype=np.float64)
+                output_data += (data - mean_data) ** 2
+                img.uncache()
+            else:
+                output_data += (img - mean_data) ** 2
+        return np.sqrt(output_data / len(input_list))
 
 
-def union(input_list):
+def union(input_list, ref_img):
     """
     union: IMGs
         Operation on binary image to keep voxels, that are non-zero, in at
         least one file.
     """
-    output_data = addition(input_list)
+    output_data = addition(input_list, ref_img)
     output_data[output_data != 0] = 1
 
     return output_data
 
 
-def intersection(input_list):
+def intersection(input_list, ref_img):
     """
     intersection: IMGs
         Operation on binary image to keep the voxels, that are non-zero,
         are present in all files.
     """
-    output_data = multiplication(input_list)
+    output_data = multiplication(input_list, ref_img)
     output_data[output_data != 0] = 1
 
     return output_data
 
 
-def difference(input_list):
+def difference(input_list, ref_img):
     """
     difference: IMG_1 IMG_2
         Operation on binary image to keep voxels from the first file that are
         not in the second file (non-zeros).
     """
     _validate_length(input_list, 2)
-    _validate_arrays(*input_list)
+    _validate_imgs(*input_list, ref_img)
 
-    output_data = copy(input_list[0]).astype(np.bool)
-    output_data[input_list[1] != 0] = 0
+    output_data = np.zeros(ref_img.header.get_data_shape(), dtype=np.float64)
+    if isinstance(input_list[0], nib.Nifti1Image):
```

Comment:
Minor issue: I feel like there are multiple instances of fetching data from either a Nifti1Image or an array itself, which could be put in a function like `load_data()` for readability

Reply:
It is because I have to uncache() right after, so I need to load it, perform the operation, unload.
Even with a wrapper function, then I need another uncache() wrapping function, it adds up to a similar amount of line.

Reply:
I see, fair enough!